Added consent creation flows

khavinshankar opened the pull request:
## Proposed Changes

- Fixes #10846
- This PR adds 2 flows to create consents:
  - Create a new consent
  - Create a consent via existing consent

@ohcnetwork/care-fe-code-reviewers

## Merge Checklist

- [ ] Add specs that demonstrate bug / test a new feature.
- [ ] Update [product documentation](https://docs.ohc.network).
- [x] Ensure that UI text is kept in I18n files.
- [ ] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [x] Request for Peer Reviews
- [] Completion of QA in Mobile Devices
- [ ] Completion of QA in Desktop Devices


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Expanded consent management with detailed descriptions and clearer messaging for various consent types and decisions.
  - Introduced an interactive dialog to easily create or link consents directly within patient encounters.
  - Enhanced file handling with context-sensitive upload options and improved document viewing customization.
  - Added a new component for managing patient consent documents, including search and preview functionalities.
  - Updated multi-factor authentication support for a more secure user experience.
  - Enhanced the PDF viewer with customizable styling options.
  
- **Bug Fixes**
  - Improved error handling messages for consent creation issues.

- **Chores**
  - Updated internal types and interfaces for better consent management and file uploads.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Jacobjeevan committed:
consent v1

khavinshankar committed:
improved existing consent linking ui: added gallery view

khavinshankar committed:
improved validations

khavinshankar committed:
added missing i18n

khavinshankar committed:
added missing translations and added consent filter by category

khavinshankar committed:
fixed file upload popup closing issue after saving

khavinshankar committed:
wired creating consent via existing consent

khavinshankar committed:
Merge branch 'develop' into issues/10846-consent

khavinshankar committed:
fixed linting and build issues

khavinshankar committed:
fixed file dialog opening issue while using existing consent

nihal467 commented:
![image](https://github.com/user-attachments/assets/da281f3b-08b6-4756-870a-19c77bfb04db)

the use existing consent is not patient specific filtered

khavinshankar commented:
@nihal467 the filtering is not happening in the backend; this pr (https://github.com/ohcnetwork/care/pull/2942) should fix that

khavinshankar committed:
Merge branch 'develop' into issues/10846-consent

khavinshankar committed:
fixed linting and type errors

khavinshankar commented on src/Utils/request/api.tsx:
```diff
@@ -16,6 +16,13 @@
 
 import { PaginatedResponse } from "@/Utils/request/types";
 import { AppointmentPatientRegister } from "@/pages/Patient/Utils";
+import {
+  ConsentListResponse,
+  ConsentModel,
+  CreateConsentRequest,
+  UpdateConsentRequest,
+  VerificationType,
+} from "@/types/consent/consent";
```
resolved

khavinshankar commented on src/Utils/request/api.tsx:
```diff
@@ -16,6 +16,13 @@
 
 import { PaginatedResponse } from "@/Utils/request/types";
 import { AppointmentPatientRegister } from "@/pages/Patient/Utils";
+import {
+  ConsentListResponse,
+  ConsentModel,
+  CreateConsentRequest,
+  UpdateConsentRequest,
+  VerificationType,
+} from "@/types/consent/consent";
```
resolved

khavinshankar committed:
Merge branch 'develop' into issues/10846-consent

nihal467 commented:
Conditionally approving it due to priority, need to work on the below-following pointer

![image](https://github.com/user-attachments/assets/2bb30a97-7a25-43f8-af0e-bcdac862af65)

- For existing consent, show the relevant detail such as the type and date of expiry etc.. so users can understand that they are usable
- Currently file name is saved for audit purpose only, but from user perspective,  they might want to edit it later,
- Add support to archive files, incase they upload some other unrelated data

![image](https://github.com/user-attachments/assets/5db7cc9e-3e6c-43f3-9815-a02ffe5b1cf1)

- Once a patient has consent, rather than showing just create a consent button, have a slide over or something to show already existing consent which is expired, ongoing, valid like that info

![image](https://github.com/user-attachments/assets/9158c084-cdc5-4c7d-8655-2af667529428)
![image](https://github.com/user-attachments/assets/44d5360b-6500-4989-88f7-c6fede299956)

- Search is not properly working when we try to type the entire file name

@khavinshankar add a proper use case along with enhancement PR for next week

CC: @gigincg

rithviknishad commented on src/types/consent/consent.ts:
```diff
@@ -0,0 +1,116 @@
+
+import { UserBase } from "@/types/user/user";
+
+export const CONSENT_CATEGORY_TYPES = [
+  {
+    id: "research",
+    label: "Research",
+    desc: "Consent to participate in research protocol and information sharing",
+  },
+  {
+    id: "patient_privacy",
+    label: "Privacy Consent",
+    desc: "Consent to collect, access, use or disclose (share) information",
+  },
+  {
+    id: "treatment",
+    label: "Treatment",
+    desc: "Consent to undergo a specific treatment",
+  },
+  {
+    id: "dnr",
+    label: "Do Not Resuscitate",
+    desc: "Consent to not receive CPR or resuscitation initiated in case of a cardiac event",
+  },
+  {
+    id: "acd",
+    label: "Advance Directive",
+    desc: "Consent given in anticipation of a potential need for medical treatment",
+  },
+  {
+    id: "adr",
+    label: "Advance Care Directive",
+    desc: "Consent for actions to be taken if they are no longer able to make decisions for themselves",
+  },
+] as const;
```
i'd suggest using i18n keys to get label and desc. from the id instead.

```
CONSENT_CATEGORY_TYPES_LABEL__research: "Research"
CONSENT_CATEGORY_TYPES_DESCRIPTION__research: "..."
```

rithviknishad commented on src/types/consent/consent.ts:
```diff
@@ -0,0 +1,116 @@
+  (category) => category.id,
+) as [(typeof CONSENT_CATEGORY_TYPES)[number]["id"]];
+
+export const CONSENT_STATUS_TYPES = [
+  { id: "active", label: "Active" },
+  { id: "inactive", label: "Inactive" },
+  { id: "draft", label: "Draft" },
+  { id: "not_done", label: "Not Done" },
+  { id: "entered_in_error", label: "Entered in Error" },
+] as const;
```
similarly here as well

rithviknishad commented on src/Utils/request/api.tsx:
```diff
@@ -573,6 +580,53 @@ const routes = {
+    },
+    list: {
+      method: "GET",
+      path: "/api/v1/patient/{patientId}/consent/",
+      TRes: Type<ConsentListResponse>(),
+    },
+    retrieve: {
+      method: "GET",
+      path: "/api/v1/patient/{patientId}/consent/{id}/",
+      TRes: Type<ConsentModel>(),
+    },
+    update: {
+      method: "PUT",
+      path: "/api/v1/patient/{patientId}/consent/{id}/",
+      TRes: Type<ConsentModel>(),
+      TBody: Type<UpdateConsentRequest>(),
+    },
+    delete: {
+      method: "DELETE",
+      path: "/api/v1/patient/{patientId}/consent/{id}/",
+    },
+    addVerification: {
+      method: "POST",
+      path: "/api/v1/patient/{patientId}/consent/{id}/add_verification/",
+      TRes: Type<ConsentModel>(),
+      TBody: Type<{ verification_type: VerificationType }>(),
+    },
+    removeVerification: {
+      method: "POST",
+      path: "/api/v1/patient/{patientId}/consent/{id}/remove_verification/",
+      TRes: Type<ConsentModel>(),
+      TBody: Type<{ verification_id: string }>(),
+    },
+    upsert: {
+      method: "POST",
+      path: "/api/v1/patient/{patientId}/consent/upsert/",
+      TRes: Type<ConsentModel>(),
+      TBody: Type<CreateConsentRequest>(),
+    },
+  },
```
new way of doing things is to have a dedicated file: `consentApi.ts` and keeping it there instead;

khavinshankar committed:
improved i18n for labels and descriptions

khavinshankar committed:
move consent apis into separate file

khavinshankar committed:
added consent sheet and removed use existing consent feature

khavinshankar commented:
@nihal467 
- Added list consent UI
- Updated the consent card UI, now shows all the details
- Removed `Use Existing Consent` feature cc: @Jacobjeevan 
- Now the consents are filterable by file name
- We won't be enabling feature to edit file name or archiving consent file but we will be adding a feature to archive consent itself in the next iteration.

nihal467 commented:
@khavinshankar add pagination to the consent as a followup PR

khavinshankar committed:
fixed invalidating query issue

Jacobjeevan commented on src/components/Consent/ConsentSheet.tsx:
```diff
@@ -0,0 +1,305 @@
+  });
+
+  const fileManager = useFileManager({
+    type: "patient",
```
```suggestion
    type: "consent",
```

khavinshankar committed:
use file type consent

khavinshankar committed:
fixed mobile responsive issue

khavinshankar committed:
hide file preview in the mobile screens

yash-learner reviewed (approved):
@khavinshankar

Just adding, currently it is possible to create a consent by uploading an invalid file type, probably we should do a file type check on clicking save

khavinshankar committed:
added view file button to show file preview in mobile view

nihal467 commented:
LGTM
